src/jacobi/_jacobi.py

```diff
@@ -1 +1,261 @@
+    if maxgrad < 0:
+        raise ValueError("maxgrad must be >= 0")
+    if step is not None:
+        if not (0 < step[0] < 0.5):
+            raise ValueError("step[0] must be between 0 and 0.5")
+        if not (0 < step[1] < 1):
+            raise ValueError("step[1] must be between 0 and 1")
+    if method is not None and method not in (-1, 0, 1):
+        raise ValueError("method must be -1, 0, 1")
+
+    x = np.asarray(x, dtype=float)
+    if mask is not None:
+        mask = np.asarray(mask)
+        if mask.dtype != bool:
+            raise ValueError("mask must be a boolean array")
+        if mask.shape != x.shape:
+            raise ValueError("mask shape must match x shape")
+
+    if diagnostic is not None:
+        diagnostic["method"] = np.zeros(x.size, dtype=np.int8)
+        diagnostic["iteration"] = np.zeros(x.size, dtype=np.uint8)
+        diagnostic["residual"] = [[] for _ in range(x.size)]
+
+    f0 = None
+    jac = None
+    err = None
+    it = np.nditer(x, flags=["c_index", "multi_index"])
+    while not it.finished:
+        k = it.index
+        kx = it.multi_index if it.has_multi_index else ...
+        if mask is not None and not mask[kx]:
+            it.iternext()
+            continue
+        xk = it[0]
+        # if step is None, use optimal step sizes for central derivatives
+        h = _steps(xk, step or (0.25, 0.5), maxiter)
+        # if method is None, auto-detect for each x[k]
+        md, f0, r = _first(method, f0, fn, x, kx, h[0], args)
+        # f0 is not guaranteed to be set here and can be still None
+
+        if md != 0 and step is None:
+            # need different step sizes for forward derivatives to avoid overlap
+            h = _steps(xk, (0.25, 0.125), maxiter)
+
+        r = np.asarray(r, dtype=float)
+        re = np.full_like(r, np.inf)
+        todo = np.ones_like(r, dtype=bool)
+        fd = [np.reshape(r.copy(), -1)]
+
+        if jac is None:  # first iteration
+            jac = np.zeros(r.shape + x.shape, dtype=r.dtype)
+            err = np.zeros(r.shape + x.shape, dtype=r.dtype)
+            if diagnostic is not None:
+                diagnostic["call"] = np.zeros((r.size, x.size), dtype=np.uint8)
+
+        if diagnostic is not None:
+            diagnostic["method"][k] = md
+            diagnostic["call"][:, k] = 2 if md == 0 else 3
+
+        for i in range(1, len(h)):
+            fdi = _derive(md, f0, fn, x, kx, h[i], args)
+            fd.append(np.reshape(fdi, -1) if i == 1 else fdi[todo])
+            if diagnostic is not None:
+                diagnostic["call"][todo, k] += 2
+                diagnostic["iteration"][k] += 1
+
+            # polynomial fit with one extra degree of freedom;
+            # use latest maxgrad + 1 data points
+            grad = min(i - 1, maxgrad)
+            start = i - (grad + 1)
+            stop = i + 1
+            q, c = np.polyfit(
+                h[start:stop] ** 2, fd[start:], grad, rcond=None, cov=True
```

Comment:
Re: 
![image](https://github.com/mdhaber/jacobi/assets/6570539/303d6e9c-745d-4d4f-abf6-29cf17cbe033)

I can see why the odd power terms are not included when the derivative is computed with central differences. Is that still true when the derivative is estimated with one-sided differences?

Reply:
That's a good point. I think you are right and this should only be done for the central derivatives. Forward and backward derivatives should use a polynomial in `h` instead of `h^2`.

Reply:
Well.. I'd think that they will require more iterations because you have twice as many error terms to cancel to get to a given order. But yes, I think that adjusting this will help.

Reply:
Yes, but on the other hand in each iteration you do half as many calculations, assuming that f0 is computed only once in the beginning. It should roughly cancel out.

Reply:
Ah, I forgot about that. I have to check, but then the h^2 powers in the polynomial fit may be correct for forward/backward formula, since it is second order like the central difference formula. At least that was the idea I had there.

Reply:
No, I don't think the order of the formula you use controls the power on `h`. It's a matter of cancelling out error terms. With central difference, there are only even-powered error terms, so you can use `h**2`. But for one-sided difference, all powers of `h` remain and need to be cancelled.

When these things are adjusted, each new function evaluation will reduce the order of the error by one whether you are using one-sided or central difference. (With central difference, you just always do two new function evaluations per iteration.)

Reply:
I'll be doing the math on this today. 

On one hand, mentally if I do Richardson extrapolation on two second order approximations that have both odd and even order error terms, I can only hope to solve for coefficients that satisfy two criteria: the first derivative itself remains, and the second order error term is cancelled. Symbolically, this would appear to leave the third order error terms if I know nothing about the coefficients.

On the other hand, I know that if I have five function evaluations, I can solve for coefficients that will give me a fourth order approximation of the derivative.

So there is definitely a way to use the `n+1` function evaluations to get an `n`th order approximation. I don't think you can get that in general by extrapolating from `n/2` second order approximations because in forming the second order approximations, you've locked some degrees of freedom you would have needed to get the higher order accuracy.

Reply:
OK, here's something a bit more concrete. I modified the Jacobi code as follows:
- Correct the sign error discussed above https://github.com/mdhaber/jacobi/pull/1#discussion_r1240208651
- Allow step sizes larger than 0.5 (so we can see convergence before running into numerical garbage).

Here's some code that allows us to plot the error as a function of the initial step size while we play with the initial step size and `maxgrad`.

<details>

```python3
f = np.exp
df = np.exp
x = 1
df_ref = df(x)
maxiter = 5
maxgrad = 0

hs = np.logspace(np.log10(1), np.log10(5), 30)
err_central = []
err_oneside = []
for h in hs:
  step = (h, 0.25)

  df_approx, _ = jacobi(f, x, maxiter=maxiter, step=step, diagonal=True, 
                        method=0, maxgrad=maxgrad)
  err_central.append(abs(df_approx - df_ref))

  df_approx, _ = jacobi(f, x, maxiter=maxiter, step=step, diagonal=True, 
                        method=1, maxgrad=maxgrad)
  err_oneside.append(abs(df_approx - df_ref))

err_central = np.array(err_central)
err_oneside = np.array(err_oneside)

order_central = np.polyfit(np.log10(hs), np.log10(err_central), 1)[0]
order_oneside = np.polyfit(np.log10(hs), np.log10(err_oneside), 1)[0]
print(f'Central difference order: {order_central}')
print(f'One-sided difference convergence order: {order_oneside}')

plt.loglog(hs, err_central, label=f'Central difference (order: {order_central})')
plt.loglog(hs, err_oneside, label=f'One-sided difference (order: {order_oneside})')
plt.legend()
plt.xlabel('Initial step size')
plt.ylabel('Error')
```

</details>


When `maxgrad=0`, we would expect both central-difference and one-sided difference to give us 2nd-order convergence.

![image](https://github.com/mdhaber/jacobi/assets/6570539/d913a01c-27e3-464e-9233-ffbde3570bcd)

Great! When `maxgrad=1`, you're combining two estimates with different step sizes. We would hope that both central-difference and one-sided difference result in a fourth-order estimate, but:
![image](https://github.com/mdhaber/jacobi/assets/6570539/a20678f3-f688-44b2-8687-0e91cf2e694c)

`maxgrad=2`
![image](https://github.com/mdhaber/jacobi/assets/6570539/77eec4c8-239e-4e4d-a62d-fbb936e152b7)

`maxgrad=3`
![image](https://github.com/mdhaber/jacobi/assets/6570539/ca537bbd-a2cb-4030-8ae7-7a6a91d179cf)

If you fix `np.polyfit` to look for a fit in `h` instead of `h**2`, you can start to see higher-order convergence, though!

Reply:
Thanks very cool, thanks for investigating and fixing this.